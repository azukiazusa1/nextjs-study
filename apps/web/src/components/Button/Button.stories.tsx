import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import Button from './Button';

export default {
  component: Button,
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args}>ボタン</Button>;

export const Default = Template.bind({});

export const isLoading = Template.bind({});
isLoading.args = {
  loading: true,
};

export const isDisabled = Template.bind({});
isDisabled.args = {
  disabled: true,
};

export const isLink = Template.bind({});
isLink.args = {
  link: true,
};

export const isOutline = Template.bind({});
isOutline.args = {
  outline: true,
};

export const isActive = Template.bind({});
isActive.args = {
  active: true,
};

const Sizes: ComponentStory<typeof Button> = (args) => (
  <div className="grid grid-cols-6 gap-4">
    <Button {...args} size="xs">
      xs
    </Button>
    <Button {...args} size="sm">
      sm
    </Button>
    <Button {...args} size="md">
      md
    </Button>
    <Button {...args} size="lg">
      lg
    </Button>
  </div>
);

export const size = Sizes.bind({});

const Colors: ComponentStory<typeof Button> = (args) => (
  <div className="grid grid-cols-6 gap-4">
    <Button {...args} color="primary">
      primary
    </Button>
    <Button {...args} color="secondary">
      secondary
    </Button>
    <Button {...args} color="success">
      success
    </Button>
    <Button {...args} color="warning">
      warning
    </Button>
    <Button {...args} color="info">
      info
    </Button>
    <Button {...args} color="error">
      error
    </Button>
    <Button {...args} color="ghost">
      ghost
    </Button>
  </div>
);

export const color = Colors.bind({});
