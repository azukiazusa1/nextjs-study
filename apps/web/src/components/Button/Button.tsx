import React from 'react';

type Props = {
  /**
   * ボタンのカラー
   * @default 'primary'
   */
  color?: 'primary' | 'secondary' | 'success' | 'warning' | 'info' | 'error' | 'ghost';
  /**
   * ボタンのサイズ
   * @default 'md'
   */
  size?: 'lg' | 'md' | 'sm' | 'xs';
  /**
   * アウトライン
   * @default false
   */
  outline?: boolean;
  /**
   * アクティブかどうか
   * @default false
   */
  active?: boolean;
  /**
   * 無効化どうか
   * @default false
   */
  disabled?: boolean;
  /**
   * リンク形式かどうか
   * @default false
   */
  link?: boolean;
  /**
   * ローディングスピナーを表示するかどうか
   * @default false
   */
  loading?: boolean;
  onClick?: () => void;
  children: React.ReactNode;
};
const Button: React.FC<Props> = ({
  color = 'primary',
  size = 'md',
  outline = false,
  active = false,
  disabled = false,
  link = false,
  loading = false,
  onClick,
  children,
}) => {
  const className = ['btn', `btn-${color}`, `btn-${size}`];

  if (outline) {
    className.push('btn-outline');
  }

  if (active) {
    className.push('active');
  }

  if (disabled) {
    className.push('disabled');
  }

  if (link) {
    className.push('btn-link');
  }

  if (loading) {
    className.push('loading');
  }

  return (
    <button type="button" className={className.join(' ')} onClick={onClick}>
      {children}
    </button>
  );
};

export default Button;
