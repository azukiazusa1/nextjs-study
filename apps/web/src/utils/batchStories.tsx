import { composeStories } from '@storybook/testing-react';
import { StoryFile } from '@storybook/testing-react/dist/types';
import { render } from '@testing-library/react';
import React from 'react';

export const batchStories = (stories: StoryFile) => {
  const testCases = Object.values(composeStories(stories)).map((Story: any) => [
    Story.storyName,
    Story,
  ]);

  test.each(testCases)('Renders %s story', async (_storyName, Story) => {
    const tree = await render(<Story />);
    expect(tree.baseElement).toMatchSnapshot();
  });
};
