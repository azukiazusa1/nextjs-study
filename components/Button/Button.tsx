const Button: React.FC = () => {
  return (
    <button type="button" className="flex text-red-400 border-2 w-24 justify-center">
      sample
    </button>
  );
};

export default Button