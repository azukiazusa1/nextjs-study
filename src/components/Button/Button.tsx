const Button: React.FC = () => {
  return (
    <button type="button" className="bg-main">
      sample
    </button>
  );
};

export default Button;
