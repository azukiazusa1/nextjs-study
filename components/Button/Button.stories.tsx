import { ComponentMeta,ComponentStory } from '@storybook/react';
import React from 'react';

import Button from './Button';

export default {
  component: Button,
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (_) => <Button />;

export const Default = Template.bind({});
