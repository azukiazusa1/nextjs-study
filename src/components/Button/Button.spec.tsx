import { batchStories } from '@/utils/batchStories';

import * as stories from './Button.stories';

describe('components/Button', () => {
  batchStories(stories);
});
