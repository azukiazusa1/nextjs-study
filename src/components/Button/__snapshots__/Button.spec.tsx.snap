// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Button Renders Default story 1`] = `
<body>
  <div>
    <button
      class="btn btn-primary btn-md"
      type="button"
    >
      ボタン
    </button>
  </div>
</body>
`;

exports[`components/Button Renders color story 1`] = `
<body>
  <div>
    <div
      class="grid grid-cols-6 gap-4"
    >
      <button
        class="btn btn-primary btn-md"
        type="button"
      >
        primary
      </button>
      <button
        class="btn btn-secondary btn-md"
        type="button"
      >
        secondary
      </button>
      <button
        class="btn btn-success btn-md"
        type="button"
      >
        success
      </button>
      <button
        class="btn btn-warning btn-md"
        type="button"
      >
        warning
      </button>
      <button
        class="btn btn-info btn-md"
        type="button"
      >
        info
      </button>
      <button
        class="btn btn-error btn-md"
        type="button"
      >
        error
      </button>
      <button
        class="btn btn-ghost btn-md"
        type="button"
      >
        ghost
      </button>
    </div>
  </div>
</body>
`;

exports[`components/Button Renders isActive story 1`] = `
<body>
  <div>
    <button
      class="btn btn-primary btn-md active"
      type="button"
    >
      ボタン
    </button>
  </div>
</body>
`;

exports[`components/Button Renders isDisabled story 1`] = `
<body>
  <div>
    <button
      class="btn btn-primary btn-md disabled"
      type="button"
    >
      ボタン
    </button>
  </div>
</body>
`;

exports[`components/Button Renders isLink story 1`] = `
<body>
  <div>
    <button
      class="btn btn-primary btn-md btn-link"
      type="button"
    >
      ボタン
    </button>
  </div>
</body>
`;

exports[`components/Button Renders isLoading story 1`] = `
<body>
  <div>
    <button
      class="btn btn-primary btn-md loading"
      type="button"
    >
      ボタン
    </button>
  </div>
</body>
`;

exports[`components/Button Renders isOutline story 1`] = `
<body>
  <div>
    <button
      class="btn btn-primary btn-md btn-outline"
      type="button"
    >
      ボタン
    </button>
  </div>
</body>
`;

exports[`components/Button Renders size story 1`] = `
<body>
  <div>
    <div
      class="grid grid-cols-6 gap-4"
    >
      <button
        class="btn btn-primary btn-xs"
        type="button"
      >
        xs
      </button>
      <button
        class="btn btn-primary btn-sm"
        type="button"
      >
        sm
      </button>
      <button
        class="btn btn-primary btn-md"
        type="button"
      >
        md
      </button>
      <button
        class="btn btn-primary btn-lg"
        type="button"
      >
        lg
      </button>
    </div>
  </div>
</body>
`;
